go-controller/pkg/ovn/policy.go

```diff
@@ -685,312 +822,288 @@ func podDeleteAllowMulticastPolicy(nbClient libovsdbclient.Client, ns string, po
+	// counters were updated, update back to initial values on error
+	defer func() {
+		if err != nil {
+			sharedPGs.updateDefaultDenyPortsCountersAdd(np, portNamesToUUIDs)
+		}
+	}()
 
-	var errs []error
+	if len(ingressDenyPorts) != 0 || len(egressDenyPorts) != 0 {
+		// db changes required
+		ops, err = libovsdbops.DeletePortsFromPortGroupOps(oc.nbClient, ops, ingressDenyPGName, ingressDenyPorts...)
+		if err != nil {
+			return fmt.Errorf("unable to get del ports from %s port group ops: %v", ingressDenyPGName, err)
+		}
 
-	ops, err := libovsdbops.AddPortsToPortGroupOps(oc.nbClient, nil, portGroupIngressDenyName, ingressDenyPorts...)
-	if err != nil {
-		oc.processLocalPodSelectorDelPods(np, obj)
-		errs = append(errs, err)
+		ops, err = libovsdbops.DeletePortsFromPortGroupOps(oc.nbClient, ops, egressDenyPGName, egressDenyPorts...)
+		if err != nil {
+			return fmt.Errorf("unable to get del ports from %s port group ops: %v", egressDenyPGName, err)
+		}
+		_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
+		if err != nil {
+			return fmt.Errorf("unable to transact del ports from default deny port groups: %v", err)
+		}
 	}
+	return nil
+}
 
-	ops, err = libovsdbops.AddPortsToPortGroupOps(oc.nbClient, ops, portGroupEgressDenyName, egressDenyPorts...)
-	if err != nil {
-		oc.processLocalPodSelectorDelPods(np, obj)
-		errs = append(errs, err)
+// handleLocalPodSelectorAddFunc adds a new pod to an existing NetworkPolicy, should be retriable.
+// ignoreErr=true should only be used for initial objects handling, relying on per-object handlers to retry.
+func (oc *Controller) handleLocalPodSelectorAddFunc(np *networkPolicy, ignoreErr bool, objs ...interface{}) error {
+	np.RLock()
+	defer np.RUnlock()
+	if np.deleted {
+		return nil
+	}
+	// get info for new pods that are not listed in np.localPods
+	portNamesToUUIDs, policyPortUUIDs, errPods := oc.getPolicyPortsForPodsAdd(np, objs...)
+	if !ignoreErr && len(errPods) > 0 {
+		pod := errPods[0].(*kapi.Pod)
+		return fmt.Errorf("unable to get port info for pod %s/%s", pod.Namespace, pod.Name)
 	}
 
-	ops, err = libovsdbops.AddPortsToPortGroupOps(oc.nbClient, ops, np.portGroupName, policyPorts...)
+	var err error
+	// add pods to policy port group
+	var ops []ovsdb.Operation
+	ops, err = libovsdbops.AddPortsToPortGroupOps(oc.nbClient, nil, np.portGroupName, policyPortUUIDs...)
 	if err != nil {
-		oc.processLocalPodSelectorDelPods(np, obj)
-		errs = append(errs, err)
+		return fmt.Errorf("unable to get ops to add new pod to policy port group: %v", err)
 	}
-
 	_, err = libovsdbops.TransactAndCheck(oc.nbClient, ops)
 	if err != nil {
-		oc.processLocalPodSelectorDelPods(np, obj)
-		errs = append(errs, err)
+		return fmt.Errorf("unable to transact add new pod to policy port group: %v", err)
+	}
+	// add pods to default deny port group
+	// make sure to only pass newly added pods
+	if err = oc.updateDenyPortGroupsAdd(np, portNamesToUUIDs); err != nil {
```

Comment:
Yeah it was written exactly with this single key=singly thread logic in mind, and I would say a lot of our code relies on that today, and probably will rely on in the future, because workqueue guarantees exactly that, and that assumption allows us to simplify the code and remove this extra locking.
In general, applying 2 events to the same object at the same has undefined result: e.g. if we add and delete object at the same time, result will depend on thread execution order which is random. So I would say for correctness we need to serialize events for the same objects anyway, then we can use this assumption to simplify objects locking.

about using set instead of port counters - yeah i considered this option, because it makes default deny pg operations idempotent, but it means we will store the same data for policy-port association 3 times.

Reply:
Regardless of worker/key isolation I still really like option 2 if only for the implicit idempotency the function gains and debugability it gives us.

Reply:
oh it looks like you already did this in your latest revision. thanks!